src/libpmemfile-posix/inode.c

```diff
@@ -0,0 +1,841 @@
+ */
+static void
+inode_map_rand_params(struct pmemfile_inode_map *c)
+{
+	/* XXX use independent random pool */
+	do {
+		c->hash_fun_a = (uint32_t)rand();
+	} while (c->hash_fun_a == 0);
+	c->hash_fun_b = (uint32_t)rand();
+}
+
+/*
+ * inode_map_alloc -- allocates inode hash map
+ */
+struct pmemfile_inode_map *
+inode_map_alloc()
+{
+	struct pmemfile_inode_map *c = calloc(1, sizeof(*c));
+
+	c->sz = 2;
+	c->buckets = calloc(1, c->sz * sizeof(c->buckets[0]));
+
+	inode_map_rand_params(c);
+	c->hash_fun_p = 32212254719ULL;
+
+	os_rwlock_init(&c->rwlock);
+
+	return c;
+}
+
+/*
+ * inode_map_free -- destroys inode hash map
+ */
+void
+inode_map_free(struct pmemfile_inode_map *c)
+{
+	for (unsigned i = 0; i < c->sz; ++i) {
+		struct inode_map_bucket *bucket = &c->buckets[i];
+
+		for (unsigned j = 0; j < BUCKET_SIZE; ++j)
+			if (bucket->arr[j].vinode)
+				FATAL("memory leak");
+	}
+
+	os_rwlock_destroy(&c->rwlock);
+	free(c->buckets);
+	free(c);
+}
+
+/*
+ * inode_hash -- returns hash value of the inode
+ */
+static inline size_t
+inode_hash(struct pmemfile_inode_map *c, TOID(struct pmemfile_inode) inode)
+{
+	return (c->hash_fun_a * inode.oid.off + c->hash_fun_b) % c->hash_fun_p;
+}
+
+/*
+ * inode_map_rebuild -- rebuilds the whole inode hash map
+ */
+static bool
+inode_map_rebuild(struct pmemfile_inode_map *c, size_t new_sz)
+{
+	struct inode_map_bucket *new_buckets =
+			calloc(1, new_sz * sizeof(new_buckets[0]));
+	size_t idx;
+
+	for (size_t i = 0; i < c->sz; ++i) {
+		struct inode_map_bucket *b = &c->buckets[i];
+
+		for (unsigned j = 0; j < BUCKET_SIZE; ++j) {
+			if (b->arr[j].pinode.oid.off == 0)
+				continue;
+
+			idx = inode_hash(c, b->arr[j].pinode) % new_sz;
+			struct inode_map_bucket *newbucket = &new_buckets[idx];
+			unsigned k;
+			for (k = 0; k < BUCKET_SIZE; ++k) {
+				if (newbucket->arr[k].pinode.oid.off == 0) {
+					newbucket->arr[k] = b->arr[j];
+					break;
+				}
+			}
+
+			if (k == BUCKET_SIZE) {
+				free(new_buckets);
+				return false;
+			}
+		}
+	}
+
+	free(c->buckets);
+	c->sz = new_sz;
+	c->buckets = new_buckets;
+
+	return true;
+}
+
+/*
+ * vinode_unregister_locked -- removes vinode from inode map
+ */
+static void
+vinode_unregister_locked(PMEMfilepool *pfp,
+		struct pmemfile_vinode *vinode)
+{
+	struct pmemfile_inode_map *c = pfp->inode_map;
+
+	size_t idx = inode_hash(c, vinode->tinode) % c->sz;
+	struct inode_map_bucket *b = &c->buckets[idx];
+	unsigned j;
+	for (j = 0; j < BUCKET_SIZE; ++j) {
+		if (b->arr[j].vinode == vinode) {
+			memset(&b->arr[j], 0, sizeof(b->arr[j]));
+			break;
+		}
+	}
+
+	if (j == BUCKET_SIZE)
+		FATAL("vinode not found");
+
+	c->inodes--;
+
+	vinode_destroy_data_state(pfp, vinode);
+
+#ifdef DEBUG
+	/* "path" field is defined only in DEBUG builds */
+	free(vinode->path);
+#endif
+	os_rwlock_destroy(&vinode->rwlock);
+	free(vinode);
+}
+
+/*
+ * _inode_get -- deals with vinode life time related to inode
```

Comment:
I would probably put a comment here stating that this function allocates a vinode if this and that, and does this and that otherwise. It was difficult to follow for me.